[litmus] a check for the DIC/IDC support

artkhyzha opened the pull request:
This draft PR is to insert into litmus7 a check whether hardware provides the DIC/IDC features.

maranget commented:
Hi @artkhyzha, I am afraid I have access to no hardware advanced enough to test this. Do you have such an access?

relokin commented on litmus/libdir/_aarch64/self.c:
```diff
@@ -43,3 +43,17 @@ static uint32_t getcachelinesize(void) {
+inline static void check_dic_idc(uint8_t need_dic, uint8_t need_idc) {
+  uint64_t ctr_el0;
+  asm volatile ("mrs %x0, ctr_el0" : "=r" (ctr_el0));
+  need_idc = need_idc? 1 : 0;
+  need_dic = need_dic? 1 : 0;
+  uint64_t test_idc = (ctr_el0 & ((uint64_t)1) << 28) ^ ((uint64_t)need_idc) << 28;
+  uint64_t test_dic = (ctr_el0 & ((uint64_t)1) << 29) ^ ((uint64_t)need_dic) << 29;
+  if (test_idc || test_dic) {
+    fprintf(stderr, "Fatal error: hardware does not support the CacheType features requested by the litmus test\n");
```
```suggestion
  bool idc = (ctr_el0 >> 28) & 1;
  bool dic = (ctr_el0 >> 29) & 1;
  if ((need_dic && !dic) || (need_idc && !idc)) {
    fprintf(stderr, "Fatal error: hardware does not support the CacheType "
            "feature IDC=%d, DIC=%d\n", need_idc, need_dic);
```

artkhyzha replied:
Thank you, this is much tidier. Applying.

relokin reviewed (changes requested):
With the comments below, I've test this on:
* A ThunderX2 which has idc=0, dic=0 and tests that require either IDC or IDC and DIC fail
* A N1SDP which has idc=1, dic=0 and tests that require IDC and DIC fail.

It would be nice if we could implement this for KVM as well but it doesn't have to be done as part of this PR. Otherwise, should we rebase and merge this?

artkhyzha commented on litmus/libdir/_aarch64/self.c:
```diff
@@ -43,3 +43,17 @@ static uint32_t getcachelinesize(void) {
+  uint64_t test_dic = (ctr_el0 & ((uint64_t)1) << 29) ^ ((uint64_t)need_dic) << 29;
+  if (test_idc || test_dic) {
+    fprintf(stderr, "Fatal error: hardware does not support the CacheType features requested by the litmus test\n");
+    fflush(stderr);
```
Just to double check: the reason why `fflush` is redundant is that `exit` flushes all buffered streams, isn't it? (And `stderr`, apparently, tends to be not buffered anyway.)

relokin replied:
Yes I think so, the next thing we do is exit, there is no value in explicitly calling `fflush()`. Keeping it won't hurt, if you prefer to keep it, but I don't see the value in it.

artkhyzha replied:
No-no, I was just asking for self-educational purposes. Thanks -- applying this suggestion too.

artkhyzha commented on litmus/libdir/_aarch64/self.c:
```diff
@@ -43,3 +43,17 @@ static uint32_t getcachelinesize(void) {
   return (1 << x) * 4 ;
 }
+
+inline static void check_dic_idc(uint8_t need_dic, uint8_t need_idc) {
```
Resolved now. It required a small tweak to `litmus/skel.ml`.

artkhyzha commented on litmus/libdir/_aarch64/self.c:
```diff
@@ -43,3 +43,17 @@ static uint32_t getcachelinesize(void) {
+
+inline static void check_dic_idc(uint8_t need_dic, uint8_t need_idc) {
+  uint64_t ctr_el0;
+  asm volatile ("mrs %x0, ctr_el0" : "=r" (ctr_el0));
```
Thanks, applied. I wonder if `litmus/libdir/_aarch64/kvm-headers.h` can benefit from the same comment.

relokin replied:
Sorry this my bad, either solution works, somehow I missed this. By default %0 will print an X register, and %x0 gives the same behavior: https://developer.arm.com/documentation/101754/0620/armclang-Reference/armclang-Inline-Assembler/Inline-assembly-template-modifiers/Template-modifiers-for-AArch64-state?lang=en

artkhyzha commented:
@relokin, thank you very much for your review. I rebased the PR against the master branch. Would you say that there are unaddressed comments blocking the merge?

P.S. I have one reservation about this PR: now that we intend to specify DIC/IDC in litmus tests globally rather than per core, one might wonder if litmus should check those specifications globally. What it does instead is checking of each allocated core's DIC/IDC support. I think that's alright and maybe even an appropriate precaution: after all, architecture provides the cache type register for each core, so why not check it.

relokin commented:
I checked once again, everything looked fine, fixed a minor glitch with the indentation, rebased and merged, thanks @artkhyzha!
